src/hinge.cpp

```diff
@@ -0,0 +1,163 @@
+#include <string>
+#include <unistd.h>
+#include <thread>
+
+class wayfire_hinge : public wf::plugin_interface_t
+{
+    enum thread_message : char
+    {
+        ENABLE_INPUT,
+        DISABLE_INPUT,
+        THREAD_EXIT,
+    };
+
+    wf::option_wrapper_t<std::string> file_name{"hinge/filename"};
+    wf::option_wrapper_t<int> poll_freq{"hinge/poll_freq"};
+    wf::option_wrapper_t<int> flip_degree{"hinge/flip_degree"};
+
+    int pipefd[2];
+    std::thread thread;
+    wl_event_source *pipe_reader;
+    bool exiting = false;
+
+    std::vector<nonstd::observer_ptr<wf::input_device_t>> get_inputs()
+    {
+        return wf::get_core().get_input_devices();
+    }
+
+    void disable_inputs()
+    {
+        for (auto inp : get_inputs())
+        {
+            auto inp_type = inp->get_wlr_handle()->type;
+            if (
+                (inp_type == wlr_input_device_type::WLR_INPUT_DEVICE_KEYBOARD) ||
+                (inp_type == wlr_input_device_type::WLR_INPUT_DEVICE_POINTER))
+            {
+                inp->set_enabled(false);
+            }
+        }
+    }
+
+    void enable_inputs()
+    {
+        for (auto inp : get_inputs())
+        {
+            inp->set_enabled();
+        }
+    }
+
+    static void setup_thread(std::string fn, int poll_freq, int flip_degree,
+        bool *exiting, int write_fd)
+    {
+        std::ifstream device_file(fn, std::ifstream::in);
+        bool input_enabled = true;
+
+        while (!*exiting)
+        {
+            char buf[4];
+            device_file.seekg(0);
+            device_file.readsome(buf, 4);
+
+            if (device_file.fail())
+            {
+                LOGE("Failed reading from hinge sensor device: ",
+                    device_file.rdstate());
+                send_message(thread_message::THREAD_EXIT, write_fd);
+                break;
+            }
+
+            int angle = std::stoi(buf);
+            if ((angle < 0) || (angle > 360))
+            {
+                LOGE("Read invalid data from hinge sensor: ", angle);
+                send_message(thread_message::THREAD_EXIT, write_fd);
+                break;
+            }
+
+            bool new_state = angle < flip_degree;
+            if (new_state != input_enabled)
+            {
+                if (new_state)
+                {
+                    send_message(thread_message::ENABLE_INPUT, write_fd);
+                } else
+                {
+                    send_message(thread_message::DISABLE_INPUT, write_fd);
+                }
+
+                input_enabled = new_state;
+            }
+
+            usleep(poll_freq * 1000); // microseconds*1000=ms
+        }
+
+        close(write_fd);
+        device_file.close();
+    }
+
+    static void send_message(thread_message message, int write_fd)
+    {
+        write(write_fd, &message, 1);
+    }
+
+    static int on_pipe_update(int fd, uint32_t mask, void *data)
+    {
+        wayfire_hinge *that = (wayfire_hinge*)data;
```

Comment:
Nitpick: Maybe `s/that/hinge/` would be better?

Reply:
As discussed on IRC, it's the base plugin class passed in as `this`, so it makes sense to refer to it as `that` here.